### ISSUE TITLE
Any plans for Processing 3.0?

### ISSUE THREAD

**Issue by blueberry** [quil/quil#166]
Processing 3.0 final has recently been released. Are there any plans for supporting it? Has anyone looked at what's new and do you have a rough estimate of how complex would be for quil to support it?

**Comment by nbeloglazov** [comment]
Same issue was recently raised in clj-processing group: https://groups.google.com/forum/#!topic/clj-processing/2OJNJUo507w The answer is yes, quil will be migrated to Processing 3. As for actual work, I haven't done anything yet. 

As for estimates I'd say it shouldn't be very hard: I don't think they changed public API much, so most function will stay unchanged. The most fun part will be in quil.applet namespace where we create PApplet and access not so public methods which probably changed. 

Personally I'm planning to do it in following way:

1. Use Processing 3.0 from pure clojure without quil to create simple empty applet. Just to see if the way to create applets from code has changed. 
2. Update quil `project.clj` to use processing 3 and try to create simple applet from quil. Fix all bugs quil.applet.
3. Run tests which exercise all publich function from quil.core and fix all broken functions, if any. Check if we need to add new functions.
4. Check that creating uberjar still works (especially with 3D). 

Any help welcome!

**Comment by blueberry** [comment]
Since I am still learning Processing, I doubt that I can provide any real help, but I will surely try to at least help with testing and reporting bugs :)
Thanks.

**Comment by blueberry** [comment]
BTW, do you have any vague plans for the *start* of the migration work. A month, the end of this year, later?

**Comment by nbeloglazov** [comment]
Started today. Progress can be tracked in https://github.com/quil/quil/tree/processing3 branch. Also to have some fun I decided to use this issue and report migration progress here. 

So I started with simple clojure program that uses raw `processing.core.PApplet`. High level API didn't change much. Here is the program: 
```clojure
(ns proc3.core
  (:import processing.core.PApplet))

(let [applet (proxy [PApplet] []
               (settings []
                 (.size this 200 200))
               (setup []
                 (.background this (int 0)))
               (draw []
                 (.stroke this 255)
                 (when (.-mousePressed this)
                   (.line this
                          (.-mouseX this) (.-mouseY this)
                          (.-pmouseX this) (.-pmouseY this))))
               (exitActual []))]

  (PApplet/runSketch (into-array String ["Bla"]) applet))
```
The main (only?) difference with processing 2 that `(.size)` method has to be called inside new `PApplet.settngs` method and not from `PApplet.setup`. 

After that created branch in quil for the migration work. Changed deps to processing 3 and fixed few bugs in `quil.Applet` due to removal `PApplet.resize`. And was able to create a simple sketch from Quil! Used only `(q/ellipse)` so sketch was very simple. Didn't test any handlers or features like :present or :resizable modes. But it looks promising: I didn't expect to get simple sketch working so soon.

**Comment by nbeloglazov** [comment]
Things to do:
- [x] make trivial sketch work
- [ ] check all functions in quil.core
- [ ] check all handler function
- [ ] check dxf and pdf renderers
- [ ] check keep-on-top
- [ ] check exit-on-close
- [ ] check resizable
- [ ] check present
- [ ] check `:display` option
- [ ] check 3d rendering
- [ ] check uberjar with 3d rendering

**Comment by nbeloglazov** [comment]
Tested handlers - they work as expected. Checked keep-on-top and resizable features. They needed migration as Processing replaced PApplet.frame (which was java class) with their own PApplet.surface w (instance of PSurface interface). Luckily it has all methods we need (setResizable and setAlwaysOnTop), so it was easy to fix. Fixed present and fullscreen mode. Now we don't need to find out size of the screen manually in quil, instead we can just call PApplet.fullScreen. But didn't test it yet with multiple screens, might see some problems.

**Comment by nbeloglazov** [comment]
Status update: 
* found out that latest processing is actually 3.0.1, not 3.0.0 and 3.0.1 is not in maven so had to fallback to hosting processing libs on clojars in quil organization;
* tested pdf and dxf, work fine. Had to move `PApplet.size()` call for pdf from `PApplet.setup` to `PApplet.settings`, but code became a little bit cleaner. 
* tested p3d renderer and packing 3d sketch to uberjars, had to create fat jars for jogl and gluegen-rt libraries as we did for previous versions of quil;
* tried to build processing dxf.jar from sources => need processing core.jar => tried to build processing to get core.jar => missing javafx on my machine => decided to upgrade to java 8 to get it => my ubuntu doesn't support java 8 => upgrade to ubuntu 15.10 => ubuntu broke while upgrading => had to reinstall it from usb flash drive => installed openjdk 8 => it doesn't have javafx => realized that I don't need build dxf.jar and can it get by downloading processing distributive from official site;

Latest development version is available on clojars: https://clojars.org/quil/versions/2.3.0-SNAPSHOT so if anyone curious - feel free to try!

**Comment by nbeloglazov** [comment]
Tried running all tests (each test exercises some function from quil.core and checks that no errors is thrown, it doesn't check that rendering is correct). Many tests failed. Some due to broken `(width)` and `(height)` function which were easy to fix. Some tests due to internal changes in processing: some functions like `PApplet.copy` can take a `PImage` as source. So some tests create a new graphics, draw on it and then pass it as `PImage` because `PGraphics` externds `PImage`. Unfortunately processing no longer supports. So we have to workaround these tests by not using graphics as PImage.

Also noticed a bug when you close sketch that uses opengl renderer (which are in the fact p2d, p3d, opengl renderers), then underlying window object doesn't get destroyed and its icon stays in OS' task bar. I found how to destroy it: by getting `GLWindow` object: `PApplet.getSurface().getNative()` and then calling `destroy()` on it. There is a small problem though, if I destroy it immediately on sketch close there is still some delayed task being executed and I get NullPointerException because window was destroyed by the time this task runs. I couldn't find a way to wait until all tasks are done and then destroy window (I'm not that familiar with how jogl library internally works). So I'm thinking about simply waiting for 0.5 sec or so and then destroy window. It's ugly, but hopefully is fine for most usecases and users shouldn't worry much unless they create/close sketches *very* fast. 

Unfortunately my work this weekend wasn't very productive, Fixed width/height, couple of tests and started fixing this destroying-windows things.

**Comment by nbeloglazov** [comment]
Tested all functions in quil.core. There was an issue with `(copy)`. The same problem I described earlier: passing `PGraphics` while `PImage` is expected. Also all tests cannot be run in single run, after a while JOGL throws error. I think it might be due to the fact that JOGL not optimized to create windows many times during single JVM run. And in our tests we have a lot (50+) tests that use JOGL renderers (:p2d or :p3d). As workaround I split all tests in sets of 50 tests in each and run single set at a time: `lein test :set-0`, `lein test :set-1`, etc. 4 sets total.
